webapp/src/modules/bid/selectors.js

```diff
@@ -0,0 +1,70 @@
+import { createSelector } from 'reselect'
+import { getAddress } from '@dapps/modules/wallet/selectors'
+import { isLoadingType } from '@dapps/modules/loading/selectors'
+
+import { BID_REQUEST, CANCEL_BID_REQUEST } from './actions'
+import { LISTING_STATUS } from 'shared/listing'
+
+export const getState = state => state.bid
+export const getData = state => getState(state).data
+export const isLoading = state => getState(state).loading.length > 0
+export const getLoading = state => getState(state).loading
+export const getError = state => getState(state).error
+
+export const isBidIdle = state => isLoadingType(getLoading(state), BID_REQUEST)
+export const isCancelIdle = state =>
+  isLoadingType(getLoading(state), CANCEL_BID_REQUEST)
+
+export const getBidById = (state, bidId, assetType, assetId) => {
+  const bids = getData(state)
+  const bid = bids[bidId]
+  return bid && bid.asset_id === assetId && bid.asset_type === assetType
+    ? bid
+    : null
+}
+
+export const getBidByAssetIdFactory = (assetType, assetId) =>
+  createSelector(getAddress, getData, (userAddress, bids) => {
+    const bidId = Object.keys(bids).filter(bidId => {
+      const bid = bids[bidId]
+      return (
+        bid &&
+        bid.asset_id === assetId &&
+        bid.asset_type === assetType &&
+        bid.bidder === userAddress
+      )
+    })[0]
+    return bidId ? bids[bidId] : null
+  })
+
+export const getBidsByAssetFactory = (isOwner, assetType, assetId) =>
+  createSelector(getAddress, getData, (userAddress, allBids) =>
+    Object.keys(allBids).reduce((bids, bidId) => {
+      const bid = allBids[bidId]
+      if (
+        bid.asset_type === assetType &&
+        bid.asset_id === assetId &&
+        bid.status === LISTING_STATUS.open &&
+        ((isOwner && bid.seller === userAddress) ||
+          (!isOwner && bid.bidder === userAddress))
+      ) {
+        return [...bids, bid]
+      }
+      return bids
+    }, [])
+  )
+
+export const getAcceptedBidsByAssetFactory = asset =>
+  createSelector(getData, allBids =>
+    Object.keys(allBids).reduce((bids, bidId) => {
+      const bid = allBids[bidId]
+      if (
+        bid.asset_type === asset.type &&
+        bid.asset_id === asset.id &&
+        bid.status === LISTING_STATUS.sold
+      ) {
+        return [...bids, bid]
+      }
+      return bids
+    }, [])
+  )
```

Comment:
**File comment 1**

I see that this check 

```js
    bid.asset_type === assetType &&
    bid.asset_id === assetId
```

is fairly common. You think it would make sense to extract and name it?

Reply:
**File comment 2**

⚠️ I see you're returning memoized selectors, but I think you're using them on mapStateToProps as " normal" selectors thus waisting the memo, i.e. `AcceptBidAssetPage.container.js`. We should either remove the createSelector or re-write the container